### ISSUE TITLE
Question about 3D covariance matrix calculation

### ISSUE THREAD

**Issue by longtermgoals** [graphdeco-inria/gaussian-splatting#762]
Hi everyone!

Amazing work, first of all, thanks for sharing this with the rest of the world!

Although I have been wondering about the implementation to calculate 3D covariance matrix using scaling and rotation matrix, code shown below (cuda_rasterizer/foward.cu in diff-gaussian-rasterization)

`
glm::mat3 M = S * R;

// Compute 3D world covariance matrix Sigma
glm::mat3 Sigma = glm::transpose(M) * M;
`

As eq (6) in the paper, covariance matrix Sigma should equal to:

Sigma = R S S^T R^T

But why in the implementation it is calculated as:

M = S * R

Sigma = M^T * M

and why this works as well.

Thank you guys so much in advance!

**Comment by jaco001** [comment]
M=SR=RS
Sigma=M S^T R^T->M (SR)^T -> M M^T -> M^T M... seems legit but who knows....

**Comment by longtermgoals** [comment]
@jaco001 , thanks for the quick reply and hint, guess I was being laze and really didn't try to figure this out, it seems obvious now thanks to your help! It can be:

```
RS = (SR)^T
Sigma = R S S^T R^T = (SR)^T SR
```

and thus the implementation, am I right?

**Comment by longtermgoals** [comment]
Hum.. I'm feeling insecure again and here is my calculations.

S is a diagonal matrix with scaling factors, and R is some 3x3 matrix. Then it should be like:

![WechatIMG1004](https://github.com/graphdeco-inria/gaussian-splatting/assets/50180943/20f88dad-4a9b-44c8-b993-9a42cf92d98d)

I'm in a little bit of doubt and not sure why the implementation didn't stick with the paper. But it really do seems working fine.

**Comment by longtermgoals** [comment]
Yeah, it's strange and not correct, my bad observation.

But it seems SR != RS either, so I'm confused again. I wonder why the implementation didn't stick to the paper, but it does seem working fine. Almighty

**Comment by altaykacan** [comment]
Hey, I think since S is a diagonal matrix `S=S^T`, and I think they work with the inverse of the rotation matrix for some efficiency reasons as mentioned [here](https://github.com/graphdeco-inria/gaussian-splatting/issues/100#issuecomment-1690378854).

So with:
- `M=S*R`
- `M^T = R^T * S^T = R^T * S`

we'd get:
`Sigma = M^T * M = R^T * S * S * R`

which basically would be the same as the formula from the paper if `R` was the inverse (inverse of a rotation matrix is its transpose) and you just replace the second `S` with the identical matrix `S^T`. 

In summary, according to my understanding, this `R` in the CUDA code is the inverse of the _R_ they use in the paper. I'm not 100% sure but I hope this helps and would appreciate any corrections or questions! :)

**Comment by longtermgoals** [comment]
> Hey, I think since S is a diagonal matrix `S=S^T`, and I think they work with the inverse of the rotation matrix for some efficiency reasons as mentioned [here](https://github.com/graphdeco-inria/gaussian-splatting/issues/100#issuecomment-1690378854).
> 
> So with:
> 
> * `M=S*R`
> * `M^T = R^T * S^T = R^T * S`
> 
> we'd get: `Sigma = M^T * M = R^T * S * S * R`
> 
> which basically would be the same as the formula from the paper if `R` was the inverse (inverse of a rotation matrix is its transpose) and you just replace the second `S` with the identical matrix `S^T`.
> 
> In summary, according to my understanding, this `R` in the CUDA code is the inverse of the _R_ they use in the paper. I'm not 100% sure but I hope this helps and would appreciate any corrections or questions! :)

Thank you so much for replying, it makes so much sense now and the issue was that I neglected the quaternions to rotation matrix in the implementation was a reverse of the R!
